[alias] Add in-place check to AliasDb and split apart hasWriters API

bwasti opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#23210 [alias] Add in-place check to AliasDb and split apart hasWriters API**

Differential Revision: [D16444529](https://our.internmc.facebook.com/intern/diff/D16444529)

bwasti committed:
[alias] Add in-place check to AliasDb

bwasti committed:
Update on "[alias] Add in-place check to AliasDb"

[wip][alias] Add in-place check to AliasDb

gh-metadata: pytorch pytorch 23210 gh/bwasti/29/head

bwasti committed:
Update on "[alias] Add in-place check to AliasDb"

[alias] Add in-place check to AliasDb

gh-metadata: pytorch pytorch 23210 gh/bwasti/29/head

bwasti committed:
Update on "[alias] Add in-place check to AliasDb"

[alias] Add in-place check to AliasDb

gh-metadata: pytorch pytorch 23210 gh/bwasti/29/head

bwasti committed:
Update on "[alias] Add in-place check to AliasDb"

[alias] Add in-place check to AliasDb

gh-metadata: pytorch pytorch 23210 gh/bwasti/29/head

suo commented on torch/csrc/jit/passes/alias_analysis.h:
```diff
@@ -61,9 +61,19 @@ class AliasDb {
 
+  // Is the operation in-place? i.e. doesn't write anywhere but locations it
+  // reads from.
+  TORCH_API bool isInPlace(Node* n) const;
```
`isMutable` might be better—given the way you have defined "in-place", there is no difference afaict

bwasti committed:
Update on "[alias] Add in-place check to AliasDb and split apart hasWriters API"

[alias] Add in-place check to AliasDb

gh-metadata: pytorch pytorch 23210 gh/bwasti/29/head
